### ISSUE TITLE
[support] Training semantics with variable-count sequence packing (greedy knapsack)

### ISSUE THREAD

**Issue by zsombor-kaiko** [NVIDIA-NeMo/Megatron-Bridge#3432]
### User problem

When using Energon's greedy knapsack packing (via select_samples_to_pack / pack_selected_samples hooks, or the upcoming `batch_level_packing=True` in PR #3323), each packed sequence contains a variable number of samples. This means the effective batch size (in samples per gradient update) fluctuates step-to-step, unlike unpacked training or fixed-MBS in-batch packing. I'm wondering if you're planning to account for this fluctuation in the hyperparameters. Tagging @DAISY-gh as the author of #3323 

### Desired outcome

Ideally, the greedy knapsack packing from Energon should increase throughput by minimizing padding tokens but should not change the training dynamics. Arguably, the easiest solution would be to use `calculate_per_token_loss=True` which might work well for pretraining but I'm wondering if that's also suitable for supervised/instruction finetuning where a sample's value might not be so directly correlated with the token count. 

### Alternatives or workarounds considered

- Estimating average packing factor upfront and decreasing global batch size (mentioned in [NeMo docs](https://docs.nvidia.com/nemo/megatron-bridge/latest/training/packed-sequences.html) but unpractical with online packing)
- Dynamic LR scaling per step based on actual samples processed. Not sure how robust this solution would be.
- Dynamic gradient accumulation frequency to scale up/down the effective batch size. This would be a nightmare with distributed training and parallelism strategies so read it as it's a very small font size. 

### Affected area

area:training

### Urgency / use case

Important but not blocking

**Comment by DAISY-gh** [comment]
> ### User problem
> When using Energon's greedy knapsack packing (via select_samples_to_pack / pack_selected_samples hooks, or the upcoming `batch_level_packing=True` in PR [#3323](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3323)), each packed sequence contains a variable number of samples. This means the effective batch size (in samples per gradient update) fluctuates step-to-step, unlike unpacked training or fixed-MBS in-batch packing. I'm wondering if you're planning to account for this fluctuation in the hyperparameters. Tagging [@DAISY-gh](https://github.com/DAISY-gh) as the author of [#3323](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3323)
> 
> ### Desired outcome
> Ideally, the greedy knapsack packing from Energon should increase throughput by minimizing padding tokens but should not change the training dynamics. Arguably, the easiest solution would be to use `calculate_per_token_loss=True` which might work well for pretraining but I'm wondering if that's also suitable for supervised/instruction finetuning where a sample's value might not be so directly correlated with the token count.
> 
> ### Alternatives or workarounds considered
> * Estimating average packing factor upfront and decreasing global batch size (mentioned in [NeMo docs](https://docs.nvidia.com/nemo/megatron-bridge/latest/training/packed-sequences.html) but unpractical with online packing)
> * Dynamic LR scaling per step based on actual samples processed. Not sure how robust this solution would be.
> * Dynamic gradient accumulation frequency to scale up/down the effective batch size. This would be a nightmare with distributed training and parallelism strategies so read it as it's a very small font size.
> 
> ### Affected area
> area:training
> 
> ### Urgency / use case
> Important but not blocking
> 
> ### Environment
> _No response_

Thanks for raising this — this is a valid concern.

When moving from BSHD-style batching to THD-style packed training, we can indeed get an implicit reweighting effect: longer sequences may contribute relatively more to the optimization signal, while shorter sequences contribute less. In other words, even if throughput improves, the training semantics can drift due to variable sample count and token composition per step.

We’ve seen related industry/workstream discussions around mitigations such as:

Loss-side correction / reweighting (e.g., normalizing or reweighting by sequence length to reduce long-sequence dominance), and
Packing policy constraints (e.g., avoiding extreme long+short co-packing patterns) to reduce per-step composition variance.
We will continue to follow similar issues and emerging solutions, and evaluate whether a robust mitigation should be added in this path.

**Comment by zsombor-kaiko** [comment]
Thanks for the quick reply @DAISY-gh! Good to hear that you are considering this issue. Could you maybe give an estimate timeline when the Energon-packing support (https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3323) will be released (with or without the mentioned mitigation strategies)?

**Comment by cuichenx** [comment]
> Arguably, the easiest solution would be to use calculate_per_token_loss=True which might work well for pretraining but I'm wondering if that's also suitable for supervised/instruction finetuning where a sample's value might not be so directly correlated with the token count.

Nvidia released the training recipes for Nemotron 3 models. In both Nemotron-3 Nano and Super THD/packed-sequence finetuning recipes, calculate_per_token_loss is turned ON. The same is true for the RL stages (stage2_rl/.../default.yaml all set calculate_per_token_loss: true).
